OAuth: clarify role & group paths prefer id_token over userinfo api (#39066)
grafana/grafana: docs/sources/auth/generic-oauth.md

Set `empty_scopes` to true to use an empty scope during authentication. By default, Grafana uses `user:email` as scope.

### Email address

> Grafana determines a user's e-mail address by querying the OAuth provider until it finds an e-mail address:

Reviewer: ```suggestion
Grafana determines a user's email address by querying the OAuth provider until it finds an e-mail address:
```

Set `empty_scopes` to true to use an empty scope during authentication. By default, Grafana uses `user:email` as scope.

> ### E-mail address

Reviewer: ```suggestion
### Email address
```

Reviewer 2: I was just following the used convention.  I personally don't mind "e-mail" as a clearer indication of the name's origin, and here in the doc as a delineation between the meaning and the config.
Would rather an e-mail spelling change separate since this isn't the subject of this PR and makes it less clear.

Reviewer: Our [style guide[(https://github.com/grafana/grafana/blob/main/contribute/style-guides/documentation-style-guide.md) says that you should follow the Google developer documentation style guide, which uses email and not e-mail.

> Grafana will attempt to determine the user's e-mail address by querying the OAuth provider as described below in the following order until an e-mail address is found:

Reviewer: ```suggestion
Grafana determines a user's e-mail address by querying the OAuth provider until it finds an e-mail address:
```

Reviewer 2: Ditto

> 1. Query the `/emails` endpoint of the OAuth provider's API (configured with `api_url`) and check for the presence of an e-mail address marked as a primary address.

Reviewer: ```suggestion
1. Query the `/emails` endpoint of the OAuth provider's API (configured with `api_url`), then check for the presence of an email address marked as a primary address.
```

Reviewer 2: ditto

> 1. If no e-mail address is found in steps (1-4), then the e-mail address of the user is set to the empty string.

Reviewer: ```suggestion
1. If no email address is found in steps (1-4), then the email address of the user is set to an empty string.
```

Reviewer 2: ditto

### Roles

> Grafana will check for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option. The JMESPath will be applied to the `id_token` first. If there is no match, the UserInfo endpoint specified via the `api_url` configuration option will be tried next. The result after evaluating the `role_attribute_path` JMESPath expression needs to be a valid Grafana role, i.e. `Viewer`, `Editor` or `Admin`.

Reviewer: ```suggestion
Grafana checks for the presence of a role using the [JMESPath](http://jmespath.org/examples.html) specified via the `role_attribute_path` configuration option. The JMESPath is applied to the `id_token` first. If there is no match, then the UserInfo endpoint specified via the `api_url` configuration option is tried next. The result after evaluation of the `role_attribute_path` JMESPath expression should be a valid Grafana role, for example, `Viewer`, `Editor` or `Admin`.
```

Reviewer 2: Was following the tense from the existing documentation - don't want to change the tense for just this section by this PR, and think that changing the tense should be a separate PR.
Re: "needs to be" - this is from the previous text and is more accurate/correct.

Reviewer 2: Don't think this should have been resolved and merged - "should" and "for example" make the documentation worse IMHO: makes it fuzzier and suggests the person setting up the integration has a choice, but it won't work if they don't set this correctly.

> See [JMESPath examples](#jmespath-examples) for more information.

Reviewer: ```suggestion
For more information, refer to the [JMESPath examples](#jmespath-examples).
```

Reviewer 2: Was existing text, but happy to take this.

### Groups / Teams

> Similarly, group mappings can be made using [JMESPath](http://jmespath.org/examples.html) with the `groups_attribute_path` configuration option. The `id_token` will be attempted first, followed by the UserInfo from the `api_url`. The result of the JMESPath expression should be a string array of groups.

Reviewer: ```suggestion
Similarly, group mappings are made using [JMESPath](http://jmespath.org/examples.html) with the `groups_attribute_path` configuration option. The `id_token` is attempted first, followed by the UserInfo from the `api_url`. The result of the JMESPath expression should be a string array of groups.
```

Reviewer 2: As above, think documentation tense change should be separate from this PR.

Furthermore, Grafana will check for the presence of at least one of the teams specified via the `team_ids` configuration option using the [JMESPath](http://jmespath.org/examples.html) specified via the `team_ids_attribute_path` configuration option. The JSON used for the path lookup is the HTTP response obtained from querying the Teams endpoint specified via the `teams_url` configuration option (using `/teams` as a fallback endpoint). The result should be a string array of Grafana Team IDs. Using this setting ensures that only certain teams is allowed to authenticate to Grafana using your OAuth provider.

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```
